### PR TITLE
exec

### PR DESCRIPTION
**Pull request by scopatz**
This should resolve all of the pending exec issues.

### CODE REVIEW

File: docs/faq.rst [hunk 1]
```diff
@@ -120,10 +120,69 @@ It is also worth noting that ``{branch_color}`` is usually the slow poke.
...
+
+    # while exec as a statement is like bash's exec
+    >>> exec gdb
+    (gdb)
+
+Yes, this is potentially confusing. This is particularly true since earlier
+versions of Python *had* an exec statement whose syntax would have clashed
+with the sh-lang command form.
+
+Yes, we are sorry. But the alternative is that import programs that use
```
Comment (astronouth7303) [inline docs/faq.rst]: "important"

Comment (scopatz) [inline docs/faq.rst]: thanks!

**Comment by laerus** [comment]
this fixes  #1302 

this doesn't seem to fix #1143 #1536 #352 (not sure if those are related)

**Comment by laerus** [comment]
either way this works as expected, `exec` and `exec()` are each what you would expect them to be, thanks @scopatz! I will play a little more with this tomorrow and merge if there is nothing more to add/change. (except if anyone wants to explore more or merge sooner)

**Comment by scopatz** [comment]
Thanks a million!

**Comment by mitnk** [comment]
Great to see this PR, sorry for my delay on it - it was hard for me to choose proper words to put into the docs ;) So glad you did it, thanks a lot!

**Comment by scopatz** [comment]
My pleasure @mitnk! Sorry it took me so long to get around to it.  I just returned home after a couple months of travel :).  Still catching up on my xonsh backlog.

**Comment by rbignon** [comment]
Hello,

There is still issues with exec, for example when you use ProxyCommand with ssh:

<pre>
$ cat ~/.ssh/config
Host owork
ProxyCommand ssh trousse -W %h:%p
$ ssh owork -v
OpenSSH_7.2p2 Ubuntu-4ubuntu2.1, OpenSSL 1.0.2g-fips  1 Mar 2016
debug1: Reading configuration data /home/rom1/.ssh/config
debug1: /home/rom1/.ssh/config line 3: Applying options for owork
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
<b>debug1: Executing proxy command: exec ssh trousse -W owork:22</b>
debug1: permanently_drop_suid: 1000
debug1: identity file /home/rom1/.ssh/id_rsa type 1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_rsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_dsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_dsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_ecdsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_ecdsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_ed25519 type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/rom1/.ssh/id_ed25519-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.1
<b>debug1: ssh_exchange_identification: &lt;function exec&gt;</b>

debug1: ssh_exchange_identification: 

ssh_exchange_identification: Connection closed by remote host
$
</pre>

If doing this command with zsh, it correctly works and the <i>debug1: ssh_exchange_identification</i> message isn't displayed.

**Comment by scopatz** [comment]
Hmm It looks like ssh is trying to run `exec` on its own without *any* arguments, which means it is probably trying to redirect command in, http://ss64.com/bash/exec.html

**Comment by rbignon** [comment]
No, that's because it launches <i>« /bin/sh /usr/local/bin/xonsh -c 'exec ssh trousse -W owork:22' »</i>, but as said in #1633, it is run as <i>« exec /usr/bin/env PYTHONUNBUFFERED=1 python3 -u -m xonsh -c exec ssh trousse -W owork:22 »</i> (without quotes), so the only command executed by xonsh is <i>exec</i> without arguments.

**Comment by PaulReiber** [comment]
Clarifying this bug/behavior:  Xonsh -c is consuming the next argument
only, not the rest of the commandline

try this:

    cd; xonsh -c 'ls /'

then:

    cd; xonsh -c ls /

The first one gives you files in / and the second one gives you files in ~

Happy to help,

Kind regards,

Paul Reiber
http://about.me/reiber

On Fri, Aug 26, 2016 at 8:21 AM, Romain Bignon <notifications@github.com>
wrote:
>
> No, that's because it launches « /bin/sh /usr/local/bin/xonsh -c 'exec
ssh trousse -W owork:22' », but as said in #1633, it is run as « exec
/usr/bin/env PYTHONUNBUFFERED=1 python3 -u -m xonsh -c exec ssh trousse -W
owork:22 » (without quotes), so the only command executed by xonsh is exec
without arguments.

**Comment by scopatz** [comment]
So will #1633 fix this when using xon.sh?

**Comment by PaulReiber** [comment]
Not sure but I don't think so.

Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by rbignon** [comment]
#1633 fixes the problem I've reported when using xon.sh.

**Pull request merged**